development_server/src/main.rs

```diff
@@ -0,0 +1,313 @@
+                        }
+                    }
+                }
+            };
+        }
+        let _ = shutdown_tx.send(());
+    });
+    let _ = component_host_server.await;
+
+    warn!("component_host_server returned");
+}
+
+pub(crate) mod protos {
+    tonic::include_proto!("development");
+
+    pub(crate) const FILE_DESCRIPTOR_SET: &[u8] =
+        tonic::include_file_descriptor_set!("development_descriptor");
+}
+
+mod rpc_server {
+    use crate::{
+        protos::{
+            development_server::Development, DeployReply, DeployRequest, EchoReply, EchoRequest,
+        },
+        ServiceCommand, ServiceCommandSink,
+    };
+    use log::info;
+    use tokio::sync::oneshot;
+    use tonic::Response;
+
+    #[derive(Debug)]
+    pub struct RpcServer {
+        command_sink: ServiceCommandSink,
+    }
+
+    impl RpcServer {
+        pub fn new(command_sink: ServiceCommandSink) -> Self {
+            Self { command_sink }
+        }
+    }
+
+    #[tonic::async_trait]
+    impl Development for RpcServer {
+        async fn echo(
+            &self,
+            request: tonic::Request<EchoRequest>,
+        ) -> Result<Response<EchoReply>, tonic::Status> {
+            Ok(Response::new(EchoReply {
+                message: request.into_inner().message,
+            }))
+        }
+
+        async fn deploy_component(
+            &self,
+            request: tonic::Request<DeployRequest>,
+        ) -> Result<Response<DeployReply>, tonic::Status> {
+            info!("received deploy_component cmd");
+            let request = request.into_inner();
+            let component_path = request.component_path;
+
+            let (reply, rx) = oneshot::channel();
+            let cmd = ServiceCommand::SwapFunctionComponent {
+                component_path,
+                reply,
+            };
+            let _ = self.command_sink.send(cmd).await;
+
+            if let Ok(Err(e)) = rx.await {
+                return Err(tonic::Status::from_error(e.into()));
+            }
+
+            Ok(Response::new(DeployReply {
+                message: "Ok".into(),
+            }))
+        }
+    }
+}
+
+mod cmd {
+    use clap::Parser;
+
+    /// Mycelia Development Server
+    #[derive(Parser, Debug)]
+    #[command(author, version, about, long_about = None)]
+    pub struct Args {
+        /// path to a function component
+        #[arg(long)]
+        pub function_component: Option<String>,
+
+        /// port rpc server should bind to
+        #[arg(long)]
+        pub rpc_port: Option<u16>,
+
+        /// port http server should bind to
+        #[arg(long)]
+        pub http_port: Option<u16>,
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    env_logger::init();
+    info!("starting up");
+
+    let args = crate::cmd::Args::parse();
+
+    let rpc_host_addr = SocketAddr::from(([127, 0, 0, 1], args.rpc_port.unwrap_or(50051)));
+    let http_host_addr = SocketAddr::from(([127, 0, 0, 1], args.http_port.unwrap_or(3001)));
```

Comment:
minor: should we check and throw if these ports are already in use by another process?

Reply:
If the ports are in use this will throw. In a follow up PR though I should add option to set the port.